cpp/src/arrow/util/decimal.cc

```diff
@@ -29,8 +29,8 @@ ARROW_EXPORT Status FromString(
   }
 
   int8_t sign = 1;
-  auto charp = s.cbegin();
-  auto end = s.cend();
+  auto charp = s.c_str();
+  auto end = charp + s.length();
```

Comment:
> after checking for s.empty()

sorry, you're right, it was hidden in "review".
Any way, if you use iterators, you have to check end before each deference (additionally to check expected character), it is not done at least thrice in #708:
52 `while (*charp == '0') {`
72, 105 `while (isdigit(*charp)) {`